multisig: key exchange update and refactor

UkoeHB opened the pull request:
## Summary
- Refactors multisig key exchange (kex) and account generation.
	- The key exchange interface has been generally simplified.
		- No more `finalize_multisig()`.
		- Most kex functionality has been sequestered in the `multisig` component library, so `wallet2` and friends only need a few relatively simple wrappers to get things done.
		- Key exchange messages contain 'round' numbers, indicating which 'round' of message exchange those message correspond to. 'Rounds' are synchronous phases of key exchange where you collect message from other participants. Recording round numbers in messages allows kex participants to more easily track the kex process.
- The multisig kex algorithm has been updated.
	- Use robust key aggregation (a.k.a. MuSig-style; see [MRL-0009](https://web.getmonero.org/resources/research-lab/pubs/MRL-0009.pdf)).
	- Validating intermediate keys during key exchange is much more robust.
	- The multisig 'base' keys from the initial kex msgs are now used as 'participant identifiers'. All messages from one participant will be signed by the same base key, including during transaction building (backward compatibility **should** be maintained with old multisig wallets).
	- Allow 1-of-N multisig wallets (M >= 2 was an arbitrary constraint).
- There are miscellaneous documentation and readability improvements related to the multisig implementation.

## TODO
- Update API/etc. docs wherever necessary (e.g. deprecate mentions of `finalize_multisig()`).

## TODO (complete)
- Manually test making multisig wallets and creating txs on stagenet.
	- Verify that the MMS still works (@rbrunner7), or update it if not.
	- **Complete**: Shoutout to @rbrunner7 for testing this.
- Manually test that backward compatibility is maintained with old multisig wallets.
	- **Complete**: Shoutout to @rbrunner7 for testing this.

## Future PRs
- Add a 'key exchange booster' function for more efficient address creation in some circumstances (very useful for 2-of-3 escrowed trades, such as those you'd expect to find in Haveno). See my [current branch](https://github.com/UkoeHB/monero/tree/multisig_addr_v2_boost).
- Move the lambda `n_choose_k()` from `cryptonote::evaluate_multisig_kex_round_msgs()` into a dedicated math library (blocked by PR #7856).

**Time spent**: About 3 - 3.5 weeks taking into account initial research.

UkoeHB commented on src/multisig/multisig.h:
```diff
@@ -28,44 +28,126 @@
+
 namespace cryptonote
 {
+  // config
+  constexpr std::uint32_t MULTISIG_MAX_SIGNERS{16};
+
```
Should I move this to `cryptonote_config.h`?

rbrunner7 reviewed (approved):
I did not look at the code changes, as I quite in general don't understand enough about the multisig implementation to see any problems there, but I did extensive testing of the branch using the MMS.

I tested 2/2 multisig, 2/3 multisig, 3/5 multisig, and 1/3 multisig wallets and made full "roundtrips": Building new wallets, funding them with a tx, syncing, and build and send out a tx. Everything worked flawlessly.

I also tested with a 2/3 wallet triplet from 2019 i.e. wallets that were not generated and configured using this new code. It was still possible to transact with them.

Good, solid work. Nice to see somebody improving multisig!

UkoeHB commented:
Updates
1. move `MAX_MULTISIG_SIGNERS` to `cryptonote_config.h`; added a comment to `multisig_kex.h` explaining why I added that limit
2. rigorously wipe variables containing user secrets that otherwise aren't exposed; added `crypto::public_key_memsafe` to `crypto.h` for that purpose (some Diffie-Hellman derivations [pubkeys] are important user secrets that need to be wiped)
3. minor cleanup thanks to @moneromooo-monero 
4. bump RPC major version to 2 (`finalize_multisig` is removed)

rbrunner7 commented:
As quite a number of code changes were made since I last tested this on [August 20](https://github.com/monero-project/monero/pull/7877#pullrequestreview-735144233), I tested again today. This time I did 2/2, 2/3 and 3/5 multisig with the help of the MMS which allows do to such tests without undue effort.

I encountered zero problems; all wallets got constructed successfully, and the out transactions were valid and all got successfully mined on testnet.

vtnerd commented on src/multisig/multisig_kex_msg.cpp:
```diff
@@ -0,0 +1,236 @@
+    }
+    else
+    {
+      for (std::size_t key_index{0}; key_index < num_keys; ++key_index)
```
If the `src/serialization` change is too problematic (because of this divergent path here), I would do a `resize` instead of a `reserve`, with a single `memcpy` for the entire vector, then loop over for the checks. It makes the loops easier as `offset` no longer needs to be updated, etc.

vtnerd commented on src/wallet/wallet2.cpp:
```diff
@@ -4983,129 +4948,126 @@ void wallet2::restore(const std::string& wallet_, const epee::wipeable_string& p
     m_account.encrypt_viewkey(chacha_key);
     m_account.decrypt_keys(chacha_key);
-    keys_reencryptor = epee::misc_utils::create_scope_leave_handler([&, this, chacha_key]() { m_account.encrypt_keys(chacha_key); m_account.decrypt_viewkey(chacha_key); });
+    keys_reencryptor = epee::misc_utils::create_scope_leave_handler(
```
Looks like a whitespace only change.

vtnerd commented on src/wallet/wallet2.cpp:
```diff
@@ -4983,129 +4948,126 @@ void wallet2::restore(const std::string& wallet_, const epee::wipeable_string& p
-      spend_skey = rct::sk2rct(cryptonote::calculate_multisig_signer_key(multisig_keys));
+  multisig_account.m_threshold = threshold;
+  multisig_account.m_signers = signers;
+  multisig_account.m_aggregate_pubkey = rct::rct2pk(rct::identity());
```
Should something like this be in the constructor? In fact several of these values could use a mandatory default.

UkoeHB replied:
Yes, if `multisig_account` were a dedicated and mature class. Making it a dedicated and mature class should be the subject of a future PR.

vtnerd replied:
Why this giant re-write without a few changes necessary to make it mature class. Its completely trivial compared to everything else I had to review for this.

vtnerd replied:
...or at the least brace initialize this class when used.

vtnerd commented on src/wallet/wallet2.cpp:
```diff
@@ -4983,129 +4948,126 @@ void wallet2::restore(const std::string& wallet_, const epee::wipeable_string& p
   m_multisig_threshold = threshold;
-  m_multisig_signers = multisig_signers;
-  ++m_multisig_rounds_passed;
+  m_multisig_signers = signers;
```
I don't see how these two values can change from above, but it may be the diff messing with me.

UkoeHB replied:
This is `make_multisig()`, so these values are uninitialized before this point.

vtnerd replied:
I thought this was setting the values in `multisig_account` again, but these are values in the `wallet2` class with the same name.

vtnerd commented on src/wallet/wallet2.cpp:
```diff
@@ -4983,129 +4948,126 @@ void wallet2::restore(const std::string& wallet_, const epee::wipeable_string& p
+    m_multisig_derivations.push_back(key_to_origins.first);
+
+  // address
+  m_account_public_address.m_spend_public_key = multisig_account.m_aggregate_pubkey;
```
This appears in the original version too - @moneromooo-monero any clue why? This would appear to be unintentional (in the original code) after a quick scan of the current version of `wallet2.cpp`.

NOTE: This same line is also below in the update.

vtnerd replied:
Yes I am asking about this assignment, and whether it should be done at all. And yes, I know it is used several times in `wallet2`, thats why I asked about this assignment.

vtnerd reviewed (commented):
FINALLY, went through all of this. Most are perf/readability related, but a few I would say a mandatory to look at. Nothing should require major re-thinking of the code - unless some issue comes to mind later (this took a long time to go through all of the logic).

UkoeHB commented on src/multisig/multisig_kex_msg.cpp:
```diff
@@ -0,0 +1,236 @@
+    }
+
+    // sets message and signing pub key
+    construct_msg(signing_privkey);
```
A `const kex_message` object is thread-safe once constructed, since all member functions are const and all member variables are private.

vtnerd replied:
I'm not recommending a change because it doesn't matter in this case - but I was suggesting a lazy/on-demand approach. It would still only be constructing once, but it would be delayed until first access (usually done with a `unique_ptr` or `boost::optional`.

UkoeHB commented on src/multisig/multisig_kex_msg.cpp:
```diff
@@ -0,0 +1,236 @@
+
+      m_msg_privkey = *(const crypto::secret_key*)(decoded.data() + offset);
+
+      CHECK_AND_ASSERT_THROW_MES(sc_check((const unsigned char*)&m_msg_privkey) == 0 &&
```
It is [here](https://github.com/monero-project/monero/blob/cc9ca953a937979c3f4cb0a332db9ac44372191f/src/crypto/crypto.h#L102).

UkoeHB commented on src/multisig/multisig_kex_msg.cpp:
```diff
@@ -0,0 +1,236 @@
+  void multisig_kex_msg::parse_and_validate_msg()
+  {
+    // check message type
+    CHECK_AND_ASSERT_THROW_MES(m_msg.size() > 0, "Kex message unexpectedly empty.");
```
I will switch the kex message serialization to `src/serialization`. This should also resolve all your comments about aliasing.

UkoeHB commented on src/multisig/multisig_kex_msg.cpp:
```diff
@@ -0,0 +1,236 @@
+    data += MULTISIG_KEX_MSG_V2_MAGIC;
+
+    // kex_round as little-endian bytes
+    for (std::size_t i{0}; i < 4; ++i)
```
yeah... I think it's a bit easier to understand with simple byte pasting

UkoeHB commented on src/wallet/wallet2.cpp:
```diff
@@ -5118,80 +5080,83 @@ std::string wallet2::make_multisig(const epee::wipeable_string &password,
-        }
-    }
+  for (const auto &msg : kex_messages)
+    expanded_msgs.emplace_back(msg);
```
Afaict wallet2 is rarely if ever exception safe. It is always the wallet2 owner who has to grab exceptions (in this case, the RPC methods).

UkoeHB commented on src/multisig/multisig.cpp:
```diff
@@ -57,134 +75,192 @@ namespace cryptonote
+
+      // convert public key (pre-merge operation)
+      // K_agg = coeff*K_base
+      rct::key converted_pubkey = rct::scalarmultKey(rct::pk2rct(key), coeff);
```
Since the `rct` library is a lot larger than `crypto`, I decided to use `rct` as much as possible. I only use `crypto` for objects that need to be memsafe (`crypto::secret_key`, `crypto::public_key_memsafe`) or if it's required by another library call.

Improving `crypto` is out of scope for this PR, and probably impractical at this point given the ubiquity of `rct`.

UkoeHB commented on src/wallet/wallet2.cpp:
```diff
@@ -5489,7 +5209,7 @@ bool wallet2::multisig(bool *ready, uint32_t *threshold, uint32_t *total) const
   return true;
 }
-
+//----------------------------------------------------------------------------------------------------
```
Since I basically re-wrote the entire multisig address implementation, I decided to do it right and clean up whitespace issues.

- `//----` barriers make it easier to see the begin/end of functions, which aids code navigation
- lambdas shouldn't be unreadable blobs of text (I am not a computer)
- `//---` barriers should be consistent (re: `multisig.cpp`, I made the barriers consistent with the new files I added)

UkoeHB commented on src/wallet/wallet2.cpp:
```diff
@@ -4983,129 +4948,126 @@ void wallet2::restore(const std::string& wallet_, const epee::wipeable_string& p
-      // N - 1 / N case
+  // sort signers
+  std::sort(signers.begin(), signers.end(),
+      [](const crypto::public_key &a, const crypto::public_key &b) -> bool
```
Const-time `memcmp` [looks harder](https://github.com/chmike/cst_time_memcmp/blob/master/consttime_memcmp.c) than it sounds.

vtnerd replied:
I think we currently use libsodium for this.

UkoeHB replied:
> Note: sodium_memcmp() is not a lexicographic comparator and is not a generic replacement for memcmp(). [source](https://libsodium.gitbook.io/doc/helpers)

vtnerd replied:
I wasn't aware of the lexicographic issue - its currently only used for `operator==` and `operator!=`.

UkoeHB commented on src/multisig/multisig.h:
```diff
@@ -28,44 +28,123 @@
+    // TODO: also record which other signers have these privkeys, to enable aggregation signing (instead of round-robin)
+    std::vector<crypto::secret_key> m_privkeys;
+    // a privkey shared by all multisig participants (e.g. private view key)
+    crypto::secret_key m_aggregate_ancillary_privkey;
```
My goal was for `multisig_account` to be conceptually isolated from its users. A multisig account has a shared private key (ancillary), and a split private key (for multisig signing).

j-berman replied:
In my super preliminary review, I have a similar comment to this effect too. Perhaps there could be more descriptive names (like `aggregate_view_privkey` instead of `aggregate_ancillary_privkey`, and `base_kex_privkey`/`base_kex_and_signing_privkey` [kinda gross] instead of `base_privkey` etc.) that achieve the same goal. The term `ancillary` relies on more context to know what it actually is, makes it a little harder to keep all the logic of what's happening/which keys are used for what in my head.

I don't have a strong opinion here/don't wanna bikeshed either, whatever you guys think is best is fine by me.

vtnerd replied:
What does that mean - it could be use for purposes other than wallet?

UkoeHB replied:
Theoretically, a multisig account can be used for any miscellaneous purpose. My point is 'separation of concerns' dictates that multisig accounts should not care about/depend on its users. I will think more about how to name things more clearly.

UkoeHB commented on src/multisig/multisig_kex.cpp:
```diff
@@ -0,0 +1,638 @@
+  //----------------------------------------------------------------------------------------------------------------------
+  // EXTERNAL
+  //----------------------------------------------------------------------------------------------------------------------
+  void multisig_kex_account_update(const std::vector<multisig_kex_msg> &expanded_msgs,
```
I decided to keep `multisig_account` as a simple struct, because turning it into a class entails a lot of extra work to develop the interface (and implement it). Right now, `wallet2` owns all its own multisig details and `multisig_account` provides a clean mapping between the two libraries. Instead, `multisig_account` should be owned and handled by `wallet2`, who doesn't know anything about the internals of `multisig_account`. We are not there yet (and this PR is already big enough).

vtnerd replied:
There isn't really any extra work - these functions become `static` or member functions. All of the static functions in the cpp would remain the same.

UkoeHB replied:
Took me a whole day, but ok...

UkoeHB commented on src/multisig/multisig_kex.cpp:
```diff
@@ -0,0 +1,638 @@
+        if (fp_result < 0)
+          return 0;
+
+        if (fp_result > std::numeric_limits<std::int32_t>::max())  // note: std::round() returns std::int32_t
```
What is this for?

vtnerd replied:
The max `std::int32_t` value can be converted into a `double` without precision loss if this is true. The rounding behavior for precision loss issues is a pain.

UkoeHB commented on src/multisig/multisig_kex.cpp:
```diff
@@ -0,0 +1,638 @@
+      // in other rounds, only the msg pubkeys are treated as msg pubkeys
+      else
+      {
+        // copy all pubkeys from message into list
```
I tried to make the messaging protocol flexible. In this case, duplicate input keys are permitted (they are ignored by the `unordered_map` and `unordered_set` use in `evaluate_multisig_kex_round_msgs()`).

vtnerd replied:
There aren't checks to: (1) ensure the remote party produced a `public_key` identical to the one produced locally, (2) that all of the messages came from the same signers, and (3) a check to ensure the aggregate pubkey is identical across all parties. Or at least, I don't recall such checks. Of all of them, (1) requires the least amount of code (now) and provides some additional sanity checks against incorrect messaging.

UkoeHB replied:
This function is 'sanitize pubkeys', it is only responsible for filtering data out of the input messages. The 'evaluate msgs' function does all these kinds of data checks.

> (1) ensure the remote party produced a `public_key` identical to the one produced locally

I explicitly don't do this test so 'round boosting' can work seamlessly (future PR). I assume you are referring to intermediate kex keys, not the final aggregate key.

Why isn't it a problem?
- If there are more rounds left, then we don't care if remote parties have the same keys as us. We take the messages from this round, remove all keys that match local keys, then make DH secrets for the next round. In the next round, we _don't care_ about our keys from the previous round.
  - If the message we make is for the last round, then we save the DH secrets as our final secret keys (who cares if other participants have these keys or not? if they lost their keys, then I guess the local signer just has more signing power than expected).
- If we are in the last round (we will make the final aggregate key from the set of messages at hand), then we already have all our private keys corresponding to signer-sets of size `N - M + 1`. All we care about is if _every set of size `N - M + 1` that excludes us shares a key_. If that condition holds, then if any set of `N - M` signers (other than the local signer) are removed, then there is always at least one person remaining with each key piece. That is the basic requirement of multisig: remove `N - M` signers, and the remaining `M` signers can always complete a signature (note: it is nonsensical from a security p.o.v. for the local signer to validate whether removing himself will allow the other `M` signers to complete a signature).

> (2) that all of the messages came from the same signers

This check is performed in `evaluate_multisig_kex_round_msgs()` where I comment `// 2 (continued). only expected signers should be recommending keys`.

> (3) a check to ensure the aggregate pubkey is identical across all parties

If there are fewer than `N - M + 1` malicious signers, then identical aggregate pubkeys between honest signers are implied, because each signer signs their kex messages. If there are `N - M + 1` malicious signers, then validating the aggregate pubkey is futile because your setup is compromised.

vtnerd replied:
I was also concerned about unintentional glitches when computing the shared secrets since there is no direct lookups to see that remote and local computations match. However, this would require that two shared secrets from one signer _accidentally_ were equivalent public keys, which is extremely unlikely. Probably about as likely as two people randomly selecting the same private key.

As you pointed out, the malicious attacker scenario isn't worth covering, because a malicious co-signer can just refuse to sign, etc., so there's no point in intentionally subverting this process.

And to be clear (I think you understand what

UkoeHB commented on src/multisig/multisig_kex.cpp:
```diff
@@ -0,0 +1,638 @@
+      // note2: avoid making temporaries that won't be memwiped
+      rct::key derivation_rct;
+      rct::scalarmultKey(derivation_rct, rct::pk2rct(pubkey_and_origins.first), rct::sk2rct(base_privkey));
+      rct::scalarmultKey(derivation_rct, derivation_rct, rct::EIGHT);
```
Why?

vtnerd replied:
An attacker could still test whether the secret key is a multiple of 2, 4 or 8 (assuming no other checks).

If `pubkey_and_origins.first` is in a subgroup, a multiplication by the subgroup order puts `derivation_rct` onto the primary group, and all other values keep it in the subgroup. So the mul8 afterward would either be putting `derivation_rct` into the primary group for the first time or be performing a mul8 in the primary group.

If the order is flipped, `derivation_rct` is in the primary group (even without other checks) before multiplying by the secret-key.

UkoeHB replied:
Curve operations are homomorphic...

Suppose `pubkey K = A + X`, where `X` is a point in the cofactor's subgroup. Let `p = privkey`.

```
p K = p A + p X
8 (p K) = 8*p A + 8*p X
            = 8*p A + p*(8 X)
            = 8*p A + p*(I)
            = 8*p A

8 K = 8 A + 8 X
       = 8 A
p (8 K) = 8*p A
```

Same thing.

vtnerd replied:
Your rebuttal is irrelevant to the issue because it doesn't address my criticism. (Assuming no other checks), this is leaking information about the secret scalar.

Or perhaps I should ask this - why is the mul8 performed at all here? Typically it is done to "clear" any torsion points on ed25519, and the comments seem to indicate that as well. This clearing should be done before the secret multiplication.

UkoeHB replied:
I am very confused. Can you show me, with pseudo-algebra, what you are talking about?

vtnerd replied:
Oops, this can be closed. Went back through relevant papers - these points I had in mind have orders of `2*l`, `4*l` or `8*l`. Its non-obvious how the torsion points interact with other points, but it makes more sense from a group theory perspective as the commutative property of multiplication still holds.

There's no check for the identity/zero point here, but is tested indirectly when trying to create new kex message.

vtnerd commented on src/multisig/multisig_kex.cpp:
```diff
@@ -0,0 +1,638 @@
+    // exclude_pubkeys should all be unique
+    for (auto it = exclude_pubkeys.begin(); it != exclude_pubkeys.end(); ++it)
+    {
+      CHECK_AND_ASSERT_THROW_MES(std::find(exclude_pubkeys.begin(), it, *it) == it,
```
Yes, correct my fault. This style was used a few lines above it, and I was referencing that.

UkoeHB commented on src/wallet/wallet_rpc_server_commands_defs.h:
```diff
@@ -46,8 +46,8 @@
-#define WALLET_RPC_VERSION_MAJOR 2
-#define WALLET_RPC_VERSION_MINOR 0
+#define WALLET_RPC_VERSION_MAJOR 1
+#define WALLET_RPC_VERSION_MINOR 24
```
Bumps minor version from 23 -> 24.

UkoeHB replied:
Note: need to revisit this when merging the PR, in case I rebase onto a higher version number.

vtnerd reviewed (commented):
Only a couple of small comments. I still have to respond to one response from the last review (I want to re-review my initial comments before responding).

UkoeHB commented on src/multisig/multisig_account.h:
```diff
@@ -0,0 +1,246 @@
+      const crypto::public_key &common_pubkey,
+      const std::uint32_t kex_rounds_complete,
+      kex_origins_map_t kex_origins_map,
+      std::string next_round_kex_message);
```
Once a kex round is complete, there is no way to recover the kex message by replaying older messages. Storing the latest message allows the operator to recover the message (e.g. if there is a communication error).

Note that older messages don't need to be stored because the 'latest' message can only be created if a full set of messages from the prior round is available (i.e. all participants already successfully completed the prior round). Round boosting violates that assumption slightly, because the boosted player will be 1 round ahead of everyone else.

vtnerd replied:
> It's not used right now, but IMO is necessary for a robust interface.

It can be added easily later.

UkoeHB replied:
Not as easily as merging it in this PR. I don't want to have to touch this code ever again.

vtnerd replied:
The original comment was definitely unrealistic - there is decent probability this code will be touched again. The edit (to the last comment) is adding unnecessary confusion in an attempt to avoid making the change, but I can't even fathom why the push back (given how small the change is) and don't really care at this point.

UkoeHB commented on src/multisig/multisig_kex_msg_serialization.h:
```diff
@@ -0,0 +1,105 @@
+  namespace serialization
+  {
+    template <class Archive>
+    inline void serialize(Archive &a, multisig::multisig_kex_msg_serializable_round1 &x, const boost::serialization::version_type ver)
```
Oh sure. Why is there both boost and in-source serialization available? Under what conditions is boost serialization used?

UkoeHB commented on src/multisig/multisig_kex_msg.cpp:
```diff
@@ -0,0 +1,266 @@
+    multisig_kex_msg_serializable_general kex_msg_general;
+
+    // try round 1 message
+    if (::serialization::serialize(b_archive_rnd1, kex_msg_rnd1))
```
For my own understanding - why wouldn't it fail?

UkoeHB commented on tests/unit_tests/multisig.cpp:
```diff
@@ -165,8 +255,88 @@ TEST(multisig, make_2_4)
+  crypto::public_key pubkey3;
+  while (!crypto::secret_key_to_public_key(rct::rct2sk(rct::skGen()), pubkey1)) {}
+  while (!crypto::secret_key_to_public_key(rct::rct2sk(rct::skGen()), pubkey2)) {}
+  while (!crypto::secret_key_to_public_key(rct::rct2sk(rct::skGen()), pubkey3)) {}
```
Update: I looked closer, and `skGen()` actually never returns 0. Pushed a small commit cleaning this up @vtnerd .

tmoravec commented:
Like in https://github.com/monero-project/monero/pull/8149, I've backported the changes to v0.17.3.0 and performed a basic smoke test:

* Create a wallet.
* Turn three wallets into 2/3 multisig.
* Receive and send transactions.
* Try to send a transaction not signed by enough participants (expected to fail)

It works :) . Thanks for the fixes!

vtnerd reviewed (approved):
Tracking changes locally; changes as expected.

h4sh3d reviewed (approved):
Second review since multisig_account has been made a class, changes looks good to me.

UkoeHB commented:
squashed
